assets/js/components/legacy-notifications/util.js

```diff
@@ -106,41 +113,36 @@ const removeDismissed = async ( notifications ) => {
  * Gets notifications from session storage, fallback to notifications API request.
  *
  * @since 1.0.0
+ * @since n.e.x.t Memoized to prevent duplicate simultaneous fetch requests from different callers.
  *
- * @return {number} Number of module notifications.
+ * @return {Promise} Object with `results` (map [slug]: notificationObject[]) and `total` (int).
  */
-export async function getModulesNotifications() {
+export const getModulesNotifications = memize( async () => {
 	const results = {};
 	let total = 0;
 
-	const modules = await modulesNotificationsToRequest();
-	const promises = [];
+	// Legacy hack: we need to use the global datastore instance here.
+	await Data.__experimentalResolveSelect( CORE_MODULES ).getModules();
+	const { isModuleActive } = Data.select( CORE_MODULES );
 
-	modules.map( async ( module ) => {
-		const promise = new Promise( async ( resolve ) => {
-			const { identifier } = module;
+	const activeModuleSlugs = modulesNotificationsToRequest().filter(
+		( slug ) => isModuleActive( slug )
+	);
```

Comment:
This was the least worst thing I could think of to fix this. In the future, we might want to expose `isModuleActive` via our `Modules` API (ie. the object provided to `registerModule`), or add support for another function arg to `registerModule` that only runs when the module is initially active (similar to `Module::register` on the PHP side).